spinalcordtoolbox/labels.py

```diff
@@ -0,0 +1,473 @@
+    """
+    og_orientation = img.orientation
+    if og_orientation != "RPI":
+        img.change_orientation("RPI")
+
+    out = zeros_like(img)
+    coordinates_input = img.getNonZeroCoordinates(sorting='z', reverse_coord=True)
+
+    # for all points with non-zeros neighbors, force the neighbors to 0
+    for i, (x, y, z, _) in enumerate(coordinates_input):
+        out.data[int(x), int(y), int(z)] = i + 1
+
+    if out.orientation != og_orientation:
+        out.change_orientation(og_orientation)
+        img.change_orientation(og_orientation)
+
+    return out
+
+
+def labelize_from_disks(img: Image, ref: Image) -> Image:
+    """
+    Create an image with regions labelized depending on values from reference.
+    Typically, user inputs a segmentation image, and labels with disks position, and this function produces
+    a segmentation image with vertebral levels labelized.
+    Labels are assumed to be non-zero and incremented from top to bottom, assuming a RPI orientation
+    :param img: segmentation
+    :param ref: reference labels
+    :returns: segmentation image with vertebral levels labelized
+    """
+    out = zeros_like(img)
+
+    coordinates_input = img.getNonZeroCoordinates()
+    coordinates_ref = ref.getNonZeroCoordinates(sorting='value')
+
+    # for all points in input, find the value that has to be set up, depending on the vertebral level
+    for x, y, z, _ in coordinates_input:
+        for j in range(len(coordinates_ref) - 1):
+            if coordinates_ref[j + 1].z < z <= coordinates_ref[j].z:
+                out.data[int(x), int(y), int(z)] = coordinates_ref[j].value
+
+    return out
+
+
+def label_vertebrae(img: Image, vertebral_levels: Sequence[int] = None) -> Image:
+    """
+    Find the center of mass of vertebral levels specified by the user.
+    :param img: source image
+    :param vertebral_levels: list of vertebral levels
+    :returns: image with labels
+    """
+
+    og_orientation = img.orientation
+    if og_orientation != "RPI":
+        img.change_orientation("RPI")
+
+    # get center of mass of each vertebral level
+    out = cubic_to_point(img)
+
+    # get list of coordinates for each label
+    list_coordinates = out.getNonZeroCoordinates(sorting='value')
+
+    # if user did not specify levels, include all:
+    if not vertebral_levels:
+        vertebral_levels = [int(i.value) for i in list_coordinates]
+
+    # loop across labels and remove those that are not listed by the user
+    for i in range(len(list_coordinates)):
+        # check if this level is NOT in vertebral_levels. if not set value to zero
+        if not vertebral_levels.count(int(list_coordinates[i].value)):
+            out.data[int(list_coordinates[i].x), int(list_coordinates[i].y), int(list_coordinates[i].z)] = 0
+
+    if out.orientation != og_orientation:
+        out.change_orientation(og_orientation)
+        img.change_orientation(og_orientation)
+
+    return out
+
+
+# FIXME [AJ]: this is slow on large images
```

Comment:
This looks fun! I love NumPy optimizations like this. LMK if you want help (or if you want me to do it.)